### PR TITLE
Use KernelConnectionMetadata instead of INotebookMetadata where possible

### PR DESCRIPTION
**Pull request by DonJayamanne**
For #4191

-   [ ] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR).
-   [ ] Title summarizes what is changing.
-   [ ] Has a [news entry](https://github.com/Microsoft/vscode-jupyter/tree/main/news) file (remember to thank yourself!).
-   [ ] Appropriate comments and documentation strings in the code.
-   [ ] Has sufficient logging.
-   [ ] Has telemetry for enhancements.
-   [ ] Unit tests & system/integration tests are added/updated.
-   [ ] [Test plan](https://github.com/Microsoft/vscode-jupyter/blob/main/.github/test_plan.md) is updated as appropriate.
-   [ ] [`package-lock.json`](https://github.com/Microsoft/vscode-jupyter/blob/main/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed).

### CODE REVIEW

File: src/client/datascience/notebookStorage/notebookModel.ts
Comment (DonJayamanne) [inline src/client/datascience/notebookStorage/notebookModel.ts:28]: In NativeNotebooks, VSC keeps track of the current selected kernel (which contains the KernelConnectionMetadata).
I didn't want to create a property in the base INotebookModel to keep track of the KernelConnection as this is only required for interactive window & WebView notebooks. 

Thus for non-native notebooks, the KernelConnectionMetadta is stored against the model.
& for NativeNotebooks we let VSC store that information.

I could create a separate class that stores this information, but then its yet another class that needs to be added to interacgive window, native notebooks & webview notebooks, I think adding it into the NotebookModel (non-native notebook model) is simpler.

Also, anyone writing code for non-native notebooks will not be able to access this property (hence guarding against accidental usage in wrong places).

File: src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts
Comment (DonJayamanne) [inline src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts:148]: Some personal TODOs: I need to check if we even need `getPreferredKernelForLocalConnection`
Its possible we need it, but i would like to understand it better (possible its only required for live share, if thats the case, i'll update comments).
FYI - I don't see when the `getPreferredKernelForLocalConnection` will get called for non-live share scenearios - hence the note

File: src/client/datascience/jupyter/liveshare/hostJupyterServer.ts
Comment (DonJayamanne) [inline src/client/datascience/jupyter/liveshare/hostJupyterServer.ts:320]: Some personal TODOs: I need to check if we even need `getPreferredKernelForLocalConnection`
Its possible we need it, but i would like to understand it better (possible its only required for live share, if thats the case, i'll update comments).
FYI - I don't see when the `getPreferredKernelForLocalConnection` will get called for non-live share scenearios - hence the note

File: src/client/datascience/notebookStorage/baseModel.ts
Comment (rchiodo) [inline src/client/datascience/notebookStorage/baseModel.ts:66]: 40 may not be enough. We might have telemetry on the number notebooks in a workspace though.

Comment (DonJayamanne) [reply src/client/datascience/notebookStorage/baseModel.ts]: I'll bump it to 100, 100 in a list is super small

Comment (DonJayamanne) [reply src/client/datascience/notebookStorage/baseModel.ts]: Updated to 100 & added telemetry to capture the size of this list

File: src/client/datascience/interactive-common/interactiveBase.ts
Comment (DonJayamanne) [inline src/client/datascience/interactive-common/interactiveBase.ts:130]: Converted these to readonly so that we can ensure we never attempt to update any properties.
Not that we have, e.g. in the case of notebook metadata there's a very specific way to update the metadata (you cannot edito this directly).
Same way with connections, added comments in the code why they are readonly (future proofing & ensure they are treated as immutable types).

File: src/client/datascience/jupyter/jupyterExecution.ts
Comment (DonJayamanne) [inline src/client/datascience/jupyter/jupyterExecution.ts:203]: An unrelated clean up, not required, but harmless & slightly better (again unrelated, couldn't help my self when fixing the typo). 😄

File: src/client/datascience/jupyter/kernels/kernel.ts
Comment (DonJayamanne) [inline src/client/datascience/jupyter/kernels/kernel.ts:139]: Got rid of the hack, yay...

File: src/client/datascience/jupyter/kernels/types.ts
Comment (DonJayamanne) [inline src/client/datascience/jupyter/kernels/types.ts:84]: Added comments to explain why i marked them readonly.

File: src/client/datascience/jupyter/liveshare/hostJupyterServer.ts
Comment (DonJayamanne) [reply src/client/datascience/jupyter/liveshare/hostJupyterServer.ts:322]: This is a WIP

File: src/test/datascience/nativeEditor.functional.test.tsx
Comment (DonJayamanne) [inline src/test/datascience/nativeEditor.functional.test.tsx:495]: Used `skip` instead of conditionally not running a test.
I was running the tests locally and spent a lot of time trying to figure out why it was workign locally & not on the server, as locally mocha results indicated it ran and completed successfully.

Found that we had if conditions that didn't use `skip`, which produces different outputs.

File: src/test/datascience/jupyter/kernels/installationPrompts.vscode.test.ts
Comment (DonJayamanne) [inline src/test/datascience/jupyter/kernels/installationPrompts.vscode.test.ts:65]: Added this as its a little difficult to figure out with all of the logging when tests start & when they end, as we have a lot of other stuff in setup & teardown (& sometimes we have additoinal logging that leaks from other tests - due to stuff not getting disposed properly)

File: src/client/datascience/interactive-common/notebookProvider.ts
Comment (DonJayamanne) [inline src/client/datascience/interactive-common/notebookProvider.ts:157]: Keeping both, if we have kernel connection, that trumps the metadata.

File: src/client/datascience/jupyter/kernels/kernelSelector.ts
Comment (DonJayamanne) [inline src/client/datascience/jupyter/kernels/kernelSelector.ts:244]: Use a better API to detect the interpreter associated with a kernel.

File: src/client/datascience/jupyter/kernels/kernelSelector.ts
Comment (DonJayamanne) [inline src/client/datascience/jupyter/kernels/kernelSelector.ts:235]: Bug fix, ensure we only do this for python kernels.

File: src/client/datascience/jupyter/kernels/kernelSelector.ts
Comment (DonJayamanne) [inline src/client/datascience/jupyter/kernels/kernelSelector.ts:273]: Use separate class to find the remote session (the kernelId is tracked against the Uri in that class, thi way we don't need to pass & store id as part of metadata, removing)

File: src/client/datascience/jupyter/kernels/kernelService.ts
Comment (DonJayamanne) [inline src/client/datascience/jupyter/kernels/kernelService.ts:312]: Fixed typo

File: src/client/datascience/jupyter/kernels/types.ts
Comment (DonJayamanne) [inline src/client/datascience/jupyter/kernels/types.ts:28]: Added comments further below why these are readonly (basically these are immutable types & connections cannot change)

File: src/client/datascience/kernel-launcher/kernelFinder.ts
Comment (DonJayamanne) [inline src/client/datascience/kernel-launcher/kernelFinder.ts:114]: Logging on CI for diagnostics, found that we don't log kernels that we have discovered.

File: src/client/datascience/kernel-launcher/kernelFinder.ts
Comment (DonJayamanne) [inline src/client/datascience/kernel-launcher/kernelFinder.ts:493]: Only for CI (logging for diagnostics)

File: .github/actions/create-venv-for-tests/action.yml
Comment (rchiodo) [inline .github/actions/create-venv-for-tests/action.yml:17]: Doesn't this install ipykernel?

File: src/client/datascience/jupyter/kernels/kernelSelections.ts
Comment (rchiodo) [inline src/client/datascience/jupyter/kernels/kernelSelections.ts:253]: Is this line necessary?

File: src/client/datascience/notebookStorage/baseModel.ts
Comment (rchiodo) [inline src/client/datascience/notebookStorage/baseModel.ts:209]: Wouldn't the id not be necessary here anymore?

File: src/client/datascience/notebookStorage/baseModel.ts
Comment (rchiodo) [inline src/client/datascience/notebookStorage/baseModel.ts:227]: I think you can get rid of this now? Is'nt the storage for this in that other location now?

File: src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts
Comment (rchiodo) [inline src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts:277]: Is this set to true anywhere?

File: .github/actions/create-venv-for-tests/action.yml
Comment (DonJayamanne) [reply .github/actions/create-venv-for-tests/action.yml:17]: & then it uninstalls it (few lines below in the script).
We install it so that we can register the kernel.

File: src/client/datascience/jupyter/kernels/kernelSelections.ts
Comment (DonJayamanne) [reply src/client/datascience/jupyter/kernels/kernelSelections.ts:253]: Yes, to update it with a new item.

File: src/client/datascience/notebookStorage/baseModel.ts
Comment (DonJayamanne) [reply src/client/datascience/notebookStorage/baseModel.ts:209]: Nope, its not. We don't need the id at all.
When connecting to remote, all we need is the Uri of the notebook and then the class will get the id based on that Uri. This way we don't need to pass this id around.

File: src/client/datascience/notebookStorage/baseModel.ts
Comment (DonJayamanne) [reply src/client/datascience/notebookStorage/baseModel.ts:227]: Yes, but the other location inherits this class. Hence is required.

File: src/client/datascience/jupyter/kernels/kernelSelections.ts
Comment (rchiodo) [reply src/client/datascience/jupyter/kernels/kernelSelections.ts:253]: Maybe I need to read up on javascript then. Isn't this a reference to the item.selection object?

```typescript
const selection = item.selection
```

So if you change selection.foo, it also changes item.selection.foo?

Comment (rchiodo) [reply src/client/datascience/jupyter/kernels/kernelSelections.ts:253]: No it does behave as I expected. Object assignment is by reference unless you explicitly clone/copy in the assignment.

File: src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts
Comment (DonJayamanne) [reply src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts:277]: Yes, in kernelProvider.ts (when getting the preferred kernel for VSCode notebooks, we don't display prompts or errors about dependencies missing). 
We do that only when running a cell.

Comment (rchiodo) [reply src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts:277]: But it's a private function? Am I missing where it's called outside of this with 'ignoreDependencyCheck' set to true?

File: src/client/datascience/notebookStorage/baseModel.ts
Comment (rchiodo) [reply src/client/datascience/notebookStorage/baseModel.ts:209]: So shouldn't you be able to remove the 'this.kernelId' from the class?

File: src/client/datascience/notebookStorage/baseModel.ts
Comment (rchiodo) [reply src/client/datascience/notebookStorage/baseModel.ts:227]: Not the _kernelConnection, the kernelId.

Comment (DonJayamanne) [reply src/client/datascience/notebookStorage/baseModel.ts]: Yup, removed, thanks

File: src/client/datascience/notebookStorage/baseModel.ts
Comment (DonJayamanne) [reply src/client/datascience/notebookStorage/baseModel.ts]: Yup removed, i thought i removed it, removed for certain now.
Also our existing functional tests will ensure this is tested correctly.

File: src/client/datascience/jupyter/kernels/kernelSelections.ts
Comment (DonJayamanne) [reply src/client/datascience/jupyter/kernels/kernelSelections.ts:253]: Oh shoot, you are right, i don't need to reset the variable, I've done that in `getKernelSelections` (few lines above, line 181).
Agreed, this is not necessary.

File: src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts
Comment (DonJayamanne) [reply src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts:277]: The argument `ignoreDependencyCheck` is passed from another public function into this private function.

Comment (DonJayamanne) [reply src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts:277]: The public function `getPreferredKernelForLocalConnection` has the same argument that gets passed into this private function.

Comment (rchiodo) [reply src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts:277]: Sorry I'm still not seeing it. The function ```installDependenciesIntoInterpreter``` you added in this PR. It's only called once as far as I can tell.

Comment (rchiodo) [reply src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts:277]: And it's called with 'false' as the value.

File: src/client/datascience/jupyter/kernels/kernel.ts
Comment (IanMatthewHuff) [inline src/client/datascience/jupyter/kernels/kernel.ts:139]: Hmm, not sure on this. But due to the name changes and the difference between INotebookMetadata and KernelConnectionMetadata we have some stuff like this:
```
metadata: undefined,
kernelConnection: this.metadata
```
Which just scans odd to me. Looks funny to not have metadata be metadata (though I get why it's not). If we have more instances like this we could rename this.metadata in this class to kernelConnection to line the names up.

File: src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts
Comment (DonJayamanne) [reply src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts:277]: In most cases its call as `false` or `undefined`. But When called from Kernel Provider in native notebooks, (which invokes `getpreferredKernelForLocalConnection`), then we pass `true`.

`false` is the default.

File: src/client/datascience/jupyter/kernels/kernel.ts
Comment (DonJayamanne) [reply src/client/datascience/jupyter/kernels/kernel.ts:139]: Agreed, will fix.

Comment (DonJayamanne) [reply src/client/datascience/jupyter/kernels/kernel.ts:139]: Fixed

File: src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts
Comment (rchiodo) [reply src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts:277]: Still don't see where 'installDependenciesIntoInterpreter ' is called more than once? What line is called on?

Comment (rchiodo) [reply src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts:277]: I believe your talking about 'installDependenciesIntoInterpreter' in kernelSelector.ts. I'm talking about the copy of that function you added to hostRawNotebookProvider. It looks to me like it's only called once. And ignoreDependencies is always false.

Comment (IanMatthewHuff) [reply src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts:277]: Yeah, I'm a bit confused on this as well. This seems like a really odd thing to be happening here in the code. Doesn't seem like a responsibility of this class.

Comment (DonJayamanne) [reply src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts:277]: > I believe your talking about 'installDependenciesIntoInterpreter' in kernelSelector.ts. I'm talking about the copy of that function you added to hostRawNotebookProvider. It looks to me like it's only called once. And ignoreDependencies is always false.

Oops. Let me check the comments again.

Comment (DonJayamanne) [reply src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts:277]: Sorry for the confusion, fixed

**Review by rchiodo — APPROVED** [review]

File: src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts
Comment (IanMatthewHuff) [reply src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts]: @DonJayamanne Sorry last comment here before approving. Why are we installing the dependencies here at this point? I remember from you and I talking about this before that we wanted dependency installation at one point. But now it's happening both in the selector and here, seems confusing and not the responsibility of this class.

Comment (DonJayamanne) [reply src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts]: I agree. However at this point there's no need to invoke the selector again as we have already selected the kernel. Happy to move this elsewhere, not sure where though?

I think we will have it at two points.
In webview & interactive window we display install instructions when selecting.
In new notebooks, we display install instructions only when starting.
As these are two separate UX  workflows I don't see how we can keep it in one place. Either way. 

Also, the code to install missting dependencies is already used in like 6 places. 
I think 1 for export, 1 for data viewer, 1 for native notebooks & the rest are just duplicates due to complex work flows.

Comment (IanMatthewHuff) [reply src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts]: You mentioned invoking the selector again. I was pretty sure that your previous changes meant that when we select we should get an interpreter with dependencies installed. What's the code path where we get here with an interpreter without dependencies installed?

Comment (IanMatthewHuff) [reply src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts]: Sorry I promise I'll stop fussing soon. Just seems like this is adding a very odd responsibility into this class. Does the kernelSelector code not run in the native notebook launch path now?

Comment (IanMatthewHuff) [reply src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts]: Bleh I don't want to hold up overnight on this as I'm not that concerned, basically just saying that I agree with your comments way back here: https://github.com/microsoft/vscode-python/pull/11341#discussion_r413392156 when you and I were debating the same issue.

**Review by IanMatthewHuff — APPROVED** [review]

File: src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts
Comment (DonJayamanne) [reply src/client/datascience/raw-kernel/liveshare/hostRawNotebookProvider.ts]: Thats ok happy to discuss.
The the past, we didn't explictly call this method in native notebooks.
In native notebooks, we call the selector & pass a flag to ignore checking dependencies.
When starting a notebook (in native notebooks), we used to call the selector once again (that's what i'm trying to remove) & at that point the flag is false (i.e. it WILL check dependnecies).
Thus in old code, when starting a native notebook, the dependencies were checked as part of the selector.

Now that we dont have the selector anymore, we still need to ensure the dependencies are still checked.

Hope that makes sense.

File: package.json
Comment (DonJayamanne) [inline package.json:1835]: Will fix before merging, trying to fix failing tests

File: src/test/datascience/notebook/notebookEditor.vscode.test.ts
Comment (DonJayamanne) [inline src/test/datascience/notebook/notebookEditor.vscode.test.ts:25]: Will fix before merging, trying to fix failing tests

**Pull request merged**
